falcon_multiqc/commands/save.py

```diff
@@ -18,159 +19,184 @@
 @click.option("-b", "--batch_description", type=click.STRING, required=False, help="Give every new batch this description.")
 @click.option("-c", "--cohort_description", type=click.STRING, required=False, help="Give every new cohort this description.")
-@click.option("-bm", "--batch_metadata", type=click.File(), required=False, help="Batch metadata file (with descriptions)")
-@click.option("-cm", "--cohort_metadata", type=click.File(), required=False, help="Cohort metadata file (with descriptions)")
-def cli(directory, sample_metadata, batch_description, cohort_description, batch_metadata, cohort_metadata):
+@click.option("-bm", "--batch_metadata", type=click.File(), required=False, help="Batch metadata file (with descriptions).")
+@click.option("-cm", "--cohort_metadata", type=click.File(), required=False, help="Cohort metadata file (with descriptions).")
+def cli(directory, sample_metadata, input_csv, batch_description, cohort_description, batch_metadata, cohort_metadata):
     """Saves the given cohort directory to the falcon_multiqc database"""
 
     with session_scope() as session:
 
-        if directory or sample_metadata:
+        # Did we get a csv?
+        if input_csv:
+
+            # Check we actually have a csv
+            if input_csv[-4:] == '.csv':
+
+                with open(input_csv, 'r') as input_csv:
+                    csv_reader = csv.reader(input_csv)
+                    header = next(csv_reader)
+
+                    # Check the headers of the csv are directory,sample_metadata
+                    if header[0] == "directory" and header[1] == "sample_metadata":
+                        with session_scope() as session:
+                            for row in csv_reader:
+                                # Check that the files in the csv actually exist
+                                if not exists(row[0]):
+                                    click.echo(f"Error: Directory {directory} does not exist."
+                                    "\nAll database entries have been rolled back, please retry after fixing")
+                                    sys.exit(1)
+                                elif not exists(row[1]):
+                                    click.echo(f"Error: Sample metadata {sample_metadata} does not exist."
+                                    "\nAll database entries have been rolled back, please retry after fixing")
+                                    sys.exit(1)
+                                else:
+                                    # save the info in that row
+                                    save_sample(row[0], row[1], session, cohort_description, batch_description)
+                    else:
+                        click.echo("CSV requires directory and sample_metadata headers.")
+                        sys.exit(1)
+            else:
+                click.echo("A csv file is required when using the --input_csv flag.")
+                sys.exit(1)
```

Comment:
@Celinet21  @Kento-Bento Are these okay to be handled like this or should they be raised as exceptions?

Reply:
These both happen before anything is actually saved to db right? If so, yeah they are fine.

Reply:
yep it looks good I think, nothing seems to be saved to the db by the time sys.exit() runs so looks good

Reply:
oki bless!